packages/compass-aggregations/src/modules/pipeline-builder/text-editor.ts

```diff
@@ -0,0 +1,225 @@
+import type { Reducer } from 'redux';
+import type { Document, MongoServerError } from 'mongodb';
+import type { PipelineBuilderThunkAction } from '..';
+import { DEFAULT_MAX_TIME_MS } from '../../constants';
+import type { PreviewOptions } from './pipeline-preview-manager';
+import {
+  DEFAULT_PREVIEW_LIMIT,
+  DEFAULT_SAMPLE_SIZE
+} from './pipeline-preview-manager';
+import { isCancelError } from '../../utils/cancellable-promise';
+import { isAction } from '../../utils/is-action';
+import type { PipelineParserError } from './pipeline-parser/utils';
+import { ActionTypes as PipelineModeActionTypes } from './pipeline-mode';
+import type { PipelineModeToggledAction } from './pipeline-mode';
+import type { PipelineBuilder } from './pipeline-builder';
+import { getStageOperator } from '../../utils/stage';
+import { CONFIRM_NEW, NEW_PIPELINE } from '../import-pipeline';
+import { RESTORE_PIPELINE } from '../saved-pipeline';
+
+export const enum EditorActionTypes {
+  EditorPreviewFetch = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetch',
+  EditorPreviewFetchSuccess = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchSuccess',
+  EditorPreviewFetchError = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchError',
+  EditorValueChange = 'compass-aggregations/pipeline-builder/text-editor/TextEditorValueChange',
+};
+
+type EditorValueChangeAction = {
+  type: EditorActionTypes.EditorValueChange;
+  pipelineText: string;
+  pipeline: Document[];
+  syntaxErrors: PipelineParserError[];
+};
+
+type EditorPreviewFetchAction = {
+  type: EditorActionTypes.EditorPreviewFetch;
+};
+
+type EditorPreviewFetchSuccessAction = {
+  type: EditorActionTypes.EditorPreviewFetchSuccess;
+  previewDocs: Document[];
+};
+
+type EditorPreviewFetchErrorAction = {
+  type: EditorActionTypes.EditorPreviewFetchError;
+  serverError: MongoServerError;
+};
+
+export type TextEditorState = {
+  pipelineText: string;
+  stageOperators: string[];
+  syntaxErrors: PipelineParserError[];
+  serverError: MongoServerError | null;
+  loading: boolean;
+  previewDocs: Document[] | null;
+};
+
+const INITIAL_STATE: TextEditorState = {
+  pipelineText: '',
+  stageOperators: [],
+  syntaxErrors: [],
+  serverError: null,
+  loading: false,
+  previewDocs: null,
+};
+
+const reducer: Reducer<TextEditorState> = (state = INITIAL_STATE, action) => {
+  if (
+    isAction<EditorValueChangeAction>(
+      action,
+      EditorActionTypes.EditorValueChange
+    ) ||
+    isAction<PipelineModeToggledAction>(
+      action,
+      PipelineModeActionTypes.PipelineModeToggled
+    ) ||
+    action.type === RESTORE_PIPELINE ||
+    action.type === CONFIRM_NEW ||
+    action.type === NEW_PIPELINE
+  ) {
+    const stageOperators = action.pipeline.length > 0
```

Comment:
See my other comment in pipeline-builder. Because you don't have an explicit value for a pipeline source that has failed to parse, you are not distinguishing correctly this case from just the empty pipeline, allowing to put UI (and builder) into a unsynced state:

![image](https://user-images.githubusercontent.com/5036933/197960253-8b9783d4-b868-429c-a069-60468bdda95d.png)